### PR TITLE
Add test for FW-CADIS based WW generation on a DAGMC model

### PR DESCRIPTION
**Pull request by shimwell**
# Description

In addition to the issue mentioned in #3490 it appears that the DAGMC geometries have a few other failure modes when used with Random Ray FW Cadis Weight Window Generation.

I couldn't find any tests running Random Ray FW Cadis Weight Window Generation with DAGMC geometry so perhaps this test can be useful

This PR is just adds a test to show a couple of failure modes.
- RuntimeError: Unknown domain type: dagmcuniverse
- RuntimeError: Maximum number of lost particles has been reached

Fixes # (issue)

# Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [ ] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by pshriwise** [comment]
I believe in this case the problem is that the bounded universe goes beyond the extent of the DAGMC universe, which has a graveyard volume and in turn has vacuum boundary conditions, which isn't valid in a universe that's not at the root level of the geometry -- hence the error message.

If this material assignment is changed to void (`'mat:void' as the DAGMC group name), the vacuum boundary conditions are then transmission and this model ran alright for me.

**Comment by pshriwise** [comment]
I've toyed around with this test locally and fond that the current error is caused by the source location being coincident with one of the tetrahedron vertices. It's something we'll want to look into more, but doesn't seem to be related to use of DAGMC models with weight windows. 

I've moved the point source in the test model off the vertex and reverted the changes commenting the surface code. Let's see how it fares in CI.

File: src/dagmc.cpp [hunk 1]
```diff
@@ -672,7 +672,7 @@ std::pair<double, int32_t> DAGCell::distance(
     p->last_dir() = u;
     p->history().reset();
   }
-  if (on_surface == 0) {
+  if (on_surface == SURFACE_NONE) {
```
Comment (shimwell) [inline src/dagmc.cpp:675]: I just noticed this change, is this the bug squashing change that solves the issue

Comment (pshriwise) [reply src/dagmc.cpp:675]: This is a readability improvement. No behavior change here. `SURFACE_NONE` is defined as `0` in `constants.h`.

https://github.com/shimwell/openmc/blob/2dbeee355cad1ce08bd6411ca92d56ac1eb3643c/include/openmc/constants.h#L371-L372

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 2]
```diff
@@ -332,3 +333,86 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+def test_ww_generation_with_dagmc():
+
+    mat1 = openmc.Material(name="no-void fuel")
+    mat1.add_nuclide("H1", 1, percent_type="ao")
+    mat1.set_density("g/cm3", 0.001)
+    mat2 = openmc.Material(name="41")
+    mat2.add_nuclide("Be9", 1, percent_type="ao")
+    mat2.set_density("g/cm3", 2.0)
+
+    my_materials = openmc.Materials([mat1, mat2])
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
    mat1 = openmc.Material(name="1")
    mat1.add_nuclide("H1", 1, percent_type="ao")
    mat1.set_density("g/cm3", 0.001)

    my_materials = openmc.Materials([mat1])
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 3]
```diff
@@ -332,3 +333,83 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+        method="fw_cadis",
+        mesh=mesh,
+        max_realizations=42,
+        particle_type='neutron',
+        energy_bounds=[0.0, 100e6]
+    )
+
+    rr_model.run()
+    # If my_geometry is set to a dagmc universe like this my_geometry = openmc.Geometry(root=dag_univ)
+    # then this line fails with RuntimeError: Unknown domain type: dagmcuniverse
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 4]
```diff
@@ -332,3 +333,83 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+    weight_windows = openmc.WeightWindowsList().from_hdf5("weight_windows.h5")
+    weight_windows = openmc.hdf5_to_wws("weight_windows.h5")
+
+    model.settings.weight_windows_on = True
+    model.settings.weight_window_checkpoints = {"collision": True, "surface": True}
+    model.settings.survival_biasing = False
+    model.settings.weight_windows = weight_windows
+
+    model.run()
+    # this line fails with RuntimeError: Maximum number of lost particles has been reached
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 5]
```diff
@@ -332,3 +333,83 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+    mat1 = openmc.Material(name="1")
+    mat1.add_nuclide("H1", 1, percent_type="ao")
+    mat1.set_density("g/cm3", 0.001)
+
+    my_materials = openmc.Materials([mat1])
+    dag_univ = openmc.DAGMCUniverse(filename=Path(__file__).parent.parent / "dagmc" / "dagmc_tetrahedral_no_graveyard.h5m")
+    bound_dag_univ = dag_univ.bounded_universe(padding_distance=10)
+    my_geometry = openmc.Geometry(root=bound_dag_univ)
+    # my_geometry = openmc.Geometry(root=dag_univ) # could set geometry like this
+    # but the ww generation then fails with RuntimeError: Unknown domain type: dagmcuniverse
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 6]
```diff
@@ -332,3 +333,78 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+        method="stochastic_slab",
+        overwrite_mgxs_library=True,
+        nparticles=20,
+        groups="CASMO-2"
+    )
+
+    rr_model.convert_to_random_ray()
+
+    mesh = openmc.RegularMesh().from_domain(rr_model)
+    mesh.dimension = (10, 10, 10)
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
    mesh.dimension = (4, 4, 4)
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 7]
```diff
@@ -332,3 +333,78 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+    mesh.id = 1
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 8]
```diff
@@ -332,3 +333,78 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+        mesh=mesh,
+        max_realizations=42,
+        particle_type='neutron',
+        energy_bounds=[0.0, 100e6]
+    )
+
+    rr_model.run()
+
+    weight_windows = openmc.WeightWindowsList().from_hdf5("weight_windows.h5")
+    weight_windows = openmc.hdf5_to_wws("weight_windows.h5")
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 9]
```diff
@@ -332,3 +333,78 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+@pytest.mark.skipif(not openmc.lib._dagmc_enabled(), reason="DAGMC CAD geometry is not enabled.")
+def test_ww_generation_with_dagmc():
+
+    mat1 = openmc.Material(name="1")
+    mat1.add_nuclide("H1", 1, percent_type="ao")
+    mat1.set_density("g/cm3", 0.001)
+
+    my_materials = openmc.Materials([mat1])
+    dag_univ = openmc.DAGMCUniverse(filename=Path(__file__).parent.parent / "dagmc" / "dagmc_tetrahedral_no_graveyard.h5m")
+    bound_dag_univ = dag_univ.bounded_universe(padding_distance=10)
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
    bound_dag_univ = dag_univ.bounded_universe(padding_distance=1)
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 10]
```diff
@@ -332,3 +333,78 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+    my_geometry = openmc.Geometry(root=bound_dag_univ)
+
+    my_settings = openmc.Settings()
+    my_settings.batches = 10
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
    my_settings.batches = 6
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 11]
```diff
@@ -332,3 +333,78 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+    my_settings.particles = 50
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
    my_settings.particles = 30
```

File: tests/unit_tests/weightwindows/test_ww_gen.py [hunk 12]
```diff
@@ -332,3 +333,78 @@ def test_ww_bounds_set_in_memory(run_in_tmpdir, model):
...
+
+    model = openmc.model.Model(my_geometry, my_materials, my_settings)
+
+    rr_model = copy.deepcopy(model)
+    rr_model.settings.inactive = 3
+
+    rr_model.convert_to_multigroup(
+        method="stochastic_slab",
+        overwrite_mgxs_library=True,
+        nparticles=20,
```
Comment (shimwell) [inline tests/unit_tests/weightwindows/test_ww_gen.py]:
```suggestion
        nparticles=10,
```
